### PR TITLE
Core framework alpha: component.ts

### PR DESCRIPTION
**Pull request by williamvinogradov**

### CODE REVIEW

File: packages/core/src/component/viewModelManager.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    });
+  };
+
+  const addViewModels = (
+    stateValue: TState,
+    subscribeToUpdates: SubscribeFunc<TState>,
+    selectorMap: SelectorMap<TState, TViewModels>,
+  ) => {
+    getKeys(selectorMap).forEach((selectorKey) => {
+      if (viewModelMap[selectorKey]) {
```
Comment (MikeVitik) [inline packages/core/src/component/viewModelManager.ts]: Looks like this code for possible future scenario. Suggest to replace it to raise error because it's impossible to understand the case of using

Comment (williamvinogradov) [reply packages/core/src/component/viewModelManager.ts]: @IlyaKhD we need your opinion here :)

File: packages/core/src/component/__test__/viewModelManager.test.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    const stateSubscribe = jest.fn();
+
+    const manager = createViewModelManager();
+    manager.addViewModels(stateValue, stateSubscribe, {
+      A: jest.fn(),
+      B: jest.fn(),
+    });
+
+    expect(createViewModelMock).toHaveBeenCalledTimes(2);
+    const [[stateArg1, subscribeArg1], [stateArg2, subscribeArg2]] = createViewModelMock.mock.calls;
```
Comment (MikeVitik) [inline packages/core/src/component/__test__/viewModelManager.test.ts]: Use toHaveBeenNthCalledWith

File: packages/core/src/component/viewModelManager.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      if (selector) {
+        viewModelMap[selectorKey] = createViewModel(stateValue, subscribeToUpdates, selector);
+      }
+    });
+  };
+
+  const getViewModels = () => viewModelMap as Readonly<ViewModelMap<TViewModels>>;
+
+  const dispose = () => {
+    getKeys(viewModelMap).forEach((key) => {
```
Comment (MikeVitik) [inline packages/core/src/component/viewModelManager.ts]: Introduce new function getValues for remove eslint comment

File: packages/core/src/__test__/viewModel.test.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
+import { createSelector, createViewModel } from '../viewModel';
```
Comment (IlyaKhD) [inline packages/core/src/__test__/viewModel.test.ts]: 🟠
I suggest that renaming files should be in a standalone PR to separate refactoring from meaningful changes

File: packages/core/src/reducer.ts [hunk 5]
```diff
@@ -1,16 +1,20 @@
...
-  TState,
+export type Reducer<
+  TState extends StateValue<ObjectType, ObjectType>,
   THandlers extends Handlers<TState>,
 > = <TAction extends keyof THandlers>(
   state: TState,
   action: TAction,
   value: Parameters<THandlers[TAction]>[1]
-) => Partial<TState>;
+) => Partial<TState['model']>;
```
Comment (IlyaKhD) [inline packages/core/src/reducer.ts]: 🔴 Reducer is a generic type and should know nothing about `TState`:

```ts
export type Reducer<
  TState,
  THandlers extends Handlers<TState, TReducedState>,
  TReducedState = TState,
> = <TAction extends keyof THandlers>(
  state: TState,
  action: TAction,
  value: Parameters<THandlers[TAction]>[1]
) => TReducedState;
```

File: packages/core/src/reducer.ts [hunk 6]
```diff
@@ -1,16 +1,20 @@
+import { StateValue } from './state';
+import { ObjectType } from './utils';
+
+export type Handlers<TState extends StateValue<ObjectType, ObjectType>> =
```
Comment (IlyaKhD) [inline packages/core/src/reducer.ts]: 🔴
```suggestion
export type Handlers<TState, TReducedState> =
```

File: packages/core/src/reducer.ts [hunk 7]
```diff
@@ -1,16 +1,20 @@
...
   THandlers extends Handlers<TState>,
 > = <TAction extends keyof THandlers>(
   state: TState,
   action: TAction,
   value: Parameters<THandlers[TAction]>[1]
-) => Partial<TState>;
+) => Partial<TState['model']>;
 
-export function createReducer<TState>() {
+export function createReducer<TState extends StateValue<ObjectType, ObjectType>>() {
```
Comment (IlyaKhD) [inline packages/core/src/reducer.ts]: 🔴
```suggestion
export function createReducer<TState, TReducedState = TState>() {
```

File: packages/core/src/viewModel.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import {
```
Comment (IlyaKhD) [inline packages/core/src/viewModel.ts]: 🟠
Please revert file renaming so that we can see what was changed

File: packages/core/src/component/__test__/viewModelManager.test.ts [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+const viewModelMock = {
+  subscribe: jest.fn(),
+  getValue: jest.fn(),
+  [DISPOSE]: jest.fn(),
+};
+
+describe('Core: Component: ViewManager', () => {
+  afterAll(() => jest.resetAllMocks());
+
+  it('add new view model', () => {
```
Comment (IlyaKhD) [inline packages/core/src/component/__test__/viewModelManager.test.ts]: 🟡
```suggestion
  it('adds new view model', () => {
```

File: packages/core/src/component/__test__/viewModelManager.test.ts [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+    const manager = createViewModelManager();
+    manager.addViewModels({}, jest.fn(), {
+      [viewModelKey]: selector,
+    });
+    const viewModels = manager.getViewModels();
+
+    expect(viewModels[viewModelKey]).toBe(expectedViewModel);
+  });
+
+  it('use the createViewModel func for adding new view models', () => {
```
Comment (IlyaKhD) [inline packages/core/src/component/__test__/viewModelManager.test.ts]: 🟡
```suggestion
  it('uses the createViewModel func for adding new view models', () => {
```

Comment (IlyaKhD) [reply packages/core/src/component/__test__/viewModelManager.test.ts]: I'm sry, but it keeps catching my eye)

File: packages/core/src/component/__test__/viewModelManager.test.ts [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+};
+
+describe('Core: Component: ViewManager', () => {
+  afterAll(() => jest.resetAllMocks());
+
+  it('add new view model', () => {
+    const expectedViewModel = {
+      ...viewModelMock,
+    };
+    const selector = jest.fn();
```
Comment (IlyaKhD) [inline packages/core/src/component/__test__/viewModelManager.test.ts]: 🟡
```suggestion
```

File: packages/core/src/component/__test__/viewModelManager.test.ts [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+    const viewModelKey = 'viewModelA';
+    createViewModelMock.mockReturnValue(expectedViewModel);
+
+    const manager = createViewModelManager();
+    manager.addViewModels({}, jest.fn(), {
+      [viewModelKey]: selector,
```
Comment (IlyaKhD) [inline packages/core/src/component/__test__/viewModelManager.test.ts]: 🟡
```suggestion
      [viewModelKey]: jest.fn(),
```

File: packages/core/src/component/__test__/viewModelManager.test.ts [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+      A: jest.fn(),
+      B: jest.fn(),
+    });
+
+    expect(createViewModelMock).toHaveBeenCalledTimes(2);
+    const [[stateArg1, subscribeArg1], [stateArg2, subscribeArg2]] = createViewModelMock.mock.calls;
+    expect(stateArg1).toBe(stateValue);
+    expect(stateArg2).toBe(stateValue);
+    expect(subscribeArg1).toBe(stateSubscribe);
+    expect(subscribeArg2).toBe(stateSubscribe);
```
Comment (IlyaKhD) [inline packages/core/src/component/__test__/viewModelManager.test.ts]: 🟠
```suggestion
    expect(createViewModelMock).toHaveBeenNthCalledWith(1, stateValue, stateSubscribe);
    expect(createViewModelMock).toHaveBeenNthCalledWith(2, stateValue, stateSubscribe);
```

File: packages/core/src/component/__test__/viewModelManager.test.ts [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+      A: undefined,
+      B: jest.fn(),
+    });
+    const viewModels = manager.getViewModels();
+
+    expect(createViewModelMock).toHaveBeenCalledTimes(1);
+    expect(Object.keys(viewModels)).toEqual(expectedViewModelKeys);
+  });
+
+  it('dispose view model before replace it with new one', () => {
```
Comment (IlyaKhD) [inline packages/core/src/component/__test__/viewModelManager.test.ts]: 🟡
```suggestion
  it('disposes view model before replace it with new one', () => {
```

File: packages/core/src/component/__test__/viewModelManager.test.ts [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    const manager = createViewModelManager();
+    manager.addViewModels({}, jest.fn(), {
+      A: firstASelector,
+    });
+    manager.addViewModels({}, jest.fn(), {
+      A: jest.fn(),
+      B: jest.fn(),
+    });
+
+    expect(testViewModel[DISPOSE]).toHaveBeenCalledTimes(1);
```
Comment (IlyaKhD) [inline packages/core/src/component/__test__/viewModelManager.test.ts]: 🟠
Both new and old view models use the same `DISPOSE` methods (because of using global variable).
Better create and test local disposeFunc:

```ts
    const testViewModel = { ...viewModelMock, [DISPOSE]: jest.fn() };
    ...
    expect(testViewModel[DISPOSE]).toHaveBeenCalledTimes(1);
```

File: packages/core/src/component/__test__/viewModelManager.test.ts [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+      C: selectorC,
+      D: jest.fn(),
+    });
+    manager.deleteViewModels('A', 'C');
+
+    expect(disposeA).toHaveBeenCalledTimes(1);
+    expect(disposeC).toHaveBeenCalledTimes(1);
+  });
+
+  it('do nothing on delete if view model with passed key doesn\'t exist', () => {
```
Comment (IlyaKhD) [inline packages/core/src/component/__test__/viewModelManager.test.ts]: 🟡
```suggestion
  it('does nothing on delete if view model with passed key doesn\'t exist', () => {
```

File: packages/core/src/component/viewModelManager.ts [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+  deleteViewModels: (...viewModelKeyArray: (keyof TViewModels)[]) => void;
+  getViewModels: () => Readonly<ViewModelMap<TViewModels>>;
+}
+
+export function createViewModelManager<TState, TViewModels extends ObjectType>()
+: Disposable<ViewModelManager<TState, TViewModels>> {
+  const viewModelMap: ViewModelMap<TViewModels> = {};
+
+  const deleteViewModels = (
+    ...viewModelKeyArray: (keyof TViewModels)[]
```
Comment (IlyaKhD) [inline packages/core/src/component/viewModelManager.ts]: 🟠 Do not use Array/Dictionary/Set words if there is no special need
```suggestion
    ...keys: (keyof TViewModels)[]
```

File: packages/core/src/component/viewModelManager.ts [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+  ): void => {
+    viewModelKeyArray.forEach((key) => {
+      viewModelMap[key]?.[DISPOSE]();
```
Comment (IlyaKhD) [inline packages/core/src/component/viewModelManager.ts]: 🟠 Hm, looks like that we should check twice or none:
```suggestion
      viewModelMap[key]?.[DISPOSE]?.();
```
or
```suggestion
      viewModelMap[key].[DISPOSE]();
```

File: packages/core/src/component/viewModelManager.ts [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+} from '../viewModel';
+
+export interface ViewModelManager<TState, TViewModels extends ObjectType> {
+  addViewModels: (
+    stateValue: TState,
+    subscribeToUpdates: SubscribeFunc<TState>,
+    selectorMap: SelectorMap<TState, TViewModels>,
+  ) => void;
+  deleteViewModels: (...viewModelKeyArray: (keyof TViewModels)[]) => void;
+  getViewModels: () => Readonly<ViewModelMap<TViewModels>>;
```
Comment (IlyaKhD) [inline packages/core/src/component/viewModelManager.ts]: 🟠 It is **ViewModel** manager already:

```suggestion
  add: (
    stateValue: TState,
    subscribeToUpdates: SubscribeFunc<TState>,
    selectorMap: SelectorMap<TState, TViewModels>,
  ) => void;
  delete: (...viewModelKeyArray: (keyof TViewModels)[]) => void;
  get: () => Readonly<ViewModelMap<TViewModels>>;
```

File: packages/core/src/component/viewModelManager.ts (hunk 1 above)
Comment (IlyaKhD) [reply packages/core/src/component/viewModelManager.ts]: It looks OK, but there is no test for it. 

Btw, should we always delete it? What if `selectorMap[selectorKey]` is undefined? A new one won't be added then (see lines below)

**Review by IlyaKhD — COMMENTED** [review]
I suggest to remove `component` folder and place these 3 files to the root. Component and it's stuff are the main part of the core package

File: packages/core/src/component/__test__/stateManager.test.ts [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+
+        const [manager] = createStateManager(stateValueMock, {}, {});
+        manager.commitUpdates();
+
+        neededCallbacks.forEach((callbackMock) => {
+          expect(callbackMock).toHaveBeenCalledTimes(1);
+        });
+      });
+
+      it('calls state\'s commitUpdates again and updates state\'s model'
```
Comment (IlyaKhD) [inline packages/core/src/component/__test__/stateManager.test.ts]: 🟡
```suggestion
      it('calls state commitUpdates again and updates statemodel'
```

File: packages/core/src/component/stateManager.ts [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    if (forceRender || (!forceRender && hasChanges)) {
+      state.triggerRender(state.getCurrent());
+    }
+
+    pendingCallbacks.forEach((callback) => callback());
+  };
+
+  // public methods.
```
Comment (IlyaKhD) [inline packages/core/src/component/stateManager.ts]: 🟠 Weird thing, let's not use comments if there is no great need
```suggestion
```

File: packages/core/src/component/stateManager.ts [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+    changeModel(newStateVersion);
+  };
+
+  return [{
+    addUpdate: state.addUpdate,
+    commitUpdates,
+    rollbackUpdates: state.rollbackUpdates,
+  }, {
+    dispatch,
+  }];
```
Comment (IlyaKhD) [inline packages/core/src/component/stateManager.ts]: 🟡 If you want to separate public from private:
```suggestion

  return [{
    addUpdate: state.addUpdate,
    rollbackUpdates: state.rollbackUpdates,
    commitUpdates() {
      state.commitUpdates();
      changeModel(state.getCurrent(), true);
    },
  }, {
      dispatch<TAction extends keyof THandlers>(
      action: TAction,
      value: Parameters<THandlers[TAction]>[1],
    ) {
      const currentStateVersion = state.getCurrent();
      const newStateVersion = {
        ...currentStateVersion,
        model: {
          ...currentStateVersion.model,
          ...reducer(currentStateVersion, action, value),
        },
      };

      changeModel(newStateVersion);
    },
  }];
```

File: packages/core/src/component/stateManager.ts [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+  stateConfig: ModelConfigMap<TModel>,
+  actionHandlers: THandlers,
+  validation: PipeFunc<StateValue<TModel, TDictionary>>[] = [],
+): StateStoreTuple<TModel, TDictionary, THandlers> {
+  const state = createState(initialState);
+  const reducer = createReducer<StateValue<TModel, TDictionary>>()(actionHandlers);
+  const validator = pipe(...validation);
+
+  // private methods.
+  const changeModel = (stateValue: StateValue<TModel, TDictionary>, forceRender = false): void => {
```
Comment (IlyaKhD) [inline packages/core/src/component/stateManager.ts]: 🔴 I believe boolean flags altering behaviour are evil seeds

```ts
  const changeModel = (stateValue: StateValue<TModel, TDictionary>): boolean => {
    ...
    return hasChanges;
  }
```

```ts
  const commitUpdates = () => {
    state.commitUpdates();
    changeModel(state.getCurrent());
    state.triggerRender(state.getCurrent());
  };
````

```ts
  const dispatch = <TAction extends keyof THandlers>(...) => {
    const hasChanges = changeModel(newStateVersion);
    if (hasChanges) {
      state.triggerRender(state.getCurrent());
    }
  };
```

File: packages/core/src/__test__/viewModel.test.ts (hunk 4 above)
Comment (williamvinogradov) [reply packages/core/src/__test__/viewModel.test.ts]: 👌

File: packages/core/src/reducer.ts (hunk 5 above)
Comment (williamvinogradov) [reply packages/core/src/reducer.ts]: Okay, agree with you.
But let's restrict TState with the Object type.

File: packages/core/src/reducer.ts (hunk 6 above)
Comment (williamvinogradov) [reply packages/core/src/reducer.ts]: Okay, agree with you.
But let's restrict TState with the Object type.

File: packages/core/src/reducer.ts (hunk 7 above)
Comment (williamvinogradov) [reply packages/core/src/reducer.ts]: Okay, agree with you.
But let's restrict TState with the Object type.

File: packages/core/src/viewModel.ts (hunk 8 above)
Comment (williamvinogradov) [reply packages/core/src/viewModel.ts]: 👌

File: packages/core/src/component/viewModelManager.ts (hunk 19 above)
Comment (williamvinogradov) [reply packages/core/src/component/viewModelManager.ts]: Agree with you, but let's not use 'delete', because it's a reserved keyword.
I suggest 'remove' instead of 'delete'.

File: packages/core/src/component/__test__/stateManager.test.ts (hunk 20 above)
Comment (williamvinogradov) [reply packages/core/src/component/__test__/stateManager.test.ts:79]: 👌

File: packages/core/src/component/stateManager.ts (hunk 21 above)
Comment (williamvinogradov) [reply packages/core/src/component/stateManager.ts]: Look like a matter of taste, but I won't ague 👌

File: packages/core/src/component/stateManager.ts (hunk 22 above)
Comment (williamvinogradov) [reply packages/core/src/component/stateManager.ts]: Let's just remove comments :)

File: packages/core/src/component/stateManager.ts (hunk 23 above)
Comment (williamvinogradov) [reply packages/core/src/component/stateManager.ts]: Fully agree with you here

File: packages/core/src/component/component.ts [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+export type CreateCoreResult<TModel extends ObjectType,
+  TDictionary extends ObjectType,
+  THandlers extends Handlers<StateValue<TModel, TDictionary>>,
+  TViewModels extends Record<PropertyKey, ObjectType>,
+  > = [
+  rootComponent: RootCoreComponent<TModel, TDictionary>,
+  containerComponent: ContainerCoreComponent<TModel, TDictionary, THandlers, TViewModels>,
+  ];
+
+export function createCoreComponent<TViewModels extends ObjectType>() {
```
Comment (IlyaKhD) [inline packages/core/src/component/component.ts]: 🟡 It's a component's core, right? not the core component?
```suggestion
export function createComponentCore<TViewModels extends ObjectType>() {
```
or even

```suggestion
export function createCore<TViewModels extends ObjectType>() {
```

File: packages/core/src/component/viewModelManager.ts (hunk 1 above)
Comment (williamvinogradov) [reply packages/core/src/component/viewModelManager.ts]: We decided to throw an Error in this case :)

File: packages/core/src/component/component.ts [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+      viewModelManager[DISPOSE]();
+    };
+
+    return [{
+      ...stateManager,
+      [DISPOSE]: dispose,
+    }, {
+      ...viewModelManager,
+      ...dispatcher,
+    }];
```
Comment (IlyaKhD) [inline packages/core/src/component/component.ts]: 🔴 Is it correct that the `DISPOSE` method belongs only to one part? What if `DISPOSE` called by the first part consumer, but the other keeps calling dispatch?
```suggestion
    return [{
      ...stateManager,
    }, {
      ...viewModelManager,
      ...dispatcher,
    },
    [DISPOSE]: dispose,
    ];
```

File: packages/core/src/component/component.ts [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+  & Dispatcher<TModel, TDictionary, THandlers>;
+
+export type CreateCoreResult<TModel extends ObjectType,
+  TDictionary extends ObjectType,
+  THandlers extends Handlers<StateValue<TModel, TDictionary>>,
+  TViewModels extends Record<PropertyKey, ObjectType>,
+  > = [
+  rootComponent: RootCoreComponent<TModel, TDictionary>,
+  containerComponent: ContainerCoreComponent<TModel, TDictionary, THandlers, TViewModels>,
+  ];
```
Comment (IlyaKhD) [inline packages/core/src/component/component.ts]: 🔴 It seems too complex to me. I prefer records to tuples, smth like:

```suggestion
export type ComponentCore<TModel extends ObjectType,
  TDictionary extends ObjectType,
  THandlers extends Handlers<StateValue<TModel, TDictionary>>,
  TViewModels extends Record<PropertyKey, ObjectType>,
> = Disposable<{
  state: StateManager<TModel, TDictionary>,
  reader: ViewModelManager<StateValue<TModel, TDictionary>, TViewModels>
    & Dispatcher<TModel, TDictionary, THandlers>,
}>;
```

which means

```ts
    return {
      state: stateManager,
      reader: {
        ...viewModelManager,
        ...dispatcher,
      },
      [DISPOSE]: dispose,
    };
```

File: packages/core/src/component/component.ts (hunk 24 above)
Comment (williamvinogradov) [reply packages/core/src/component/component.ts]: Let's call it createCore

File: packages/core/src/component/index.ts [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+  & Dispatcher<TModel, TDictionary, THandlers>;
+
+export type CreateCoreResult<TModel extends ObjectType,
+  TDictionary extends ObjectType,
+  THandlers extends Handlers<StateValue<TModel, TDictionary>>,
+  TViewModels extends Record<PropertyKey, ObjectType>,
+  > = [
+  rootComponent: RootCore<TModel, TDictionary>,
+  containerComponent: ContainerCore<TModel, TDictionary, THandlers, TViewModels>,
+  ];
```
Comment (IlyaKhD) [inline packages/core/src/component/index.ts]:
```suggestion
export type CoreState<TModel extends ObjectType, TDictionary extends ObjectType> =
  Disposable<StateManager<TModel, TDictionary>>;

export type CoreStateReader<
  TModel extends ObjectType,
  TDictionary extends ObjectType,
  THandlers extends Handlers<StateValue<TModel, TDictionary>>,
  TViewModels extends ObjectType,
  > = ViewModelManager<StateValue<TModel, TDictionary>, TViewModels>
  & Dispatcher<TModel, TDictionary, THandlers>;

export type CreateCoreResult<TModel extends ObjectType,
  TDictionary extends ObjectType,
  THandlers extends Handlers<StateValue<TModel, TDictionary>>,
  TViewModels extends Record<PropertyKey, ObjectType>,
  > = [
  state: RootCore<TModel, TDictionary>,
  stateReader: ContainerCore<TModel, TDictionary, THandlers, TViewModels>,
  ];
```

Comment (IlyaKhD) [reply packages/core/src/component/index.ts]: These are not components.... I like 'reader' word here, not sure about 'state'

**Review by MikeVitik — APPROVED** [review]

**Review by IlyaKhD — APPROVED** [review]
